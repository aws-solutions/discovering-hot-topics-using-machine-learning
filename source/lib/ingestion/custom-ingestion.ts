#!/usr/bin/env node
/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import * as cdk from 'aws-cdk-lib';
import * as events from 'aws-cdk-lib/aws-events';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as kinesis from 'aws-cdk-lib/aws-kinesis';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as s3 from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';
import * as fs from 'fs';
import { S3ToEventBridgeToLambda } from '../s3-event-notification/s3-eventbridge-lambda';
import path = require('path');

export class CustomIngestion extends cdk.NestedStack {
    public readonly s3Bucket: s3.Bucket;

    constructor(scope: Construct, id: string, props: cdk.NestedStackProps) {
        super(scope, id, props);

        const _streamArn = new cdk.CfnParameter(this, 'StreamARN', {
            type: 'String',
            description:
                'The name of the stream where the records from custom ingestion should be published for analysis',
            allowedPattern: '^arn:\\S+:kinesis:\\S+:\\d{12}:stream/\\S+$',
            constraintDescription: 'Please provide the Amazon Kinesis Data Streams name'
        });

        const _accessLoggingBucket = new cdk.CfnParameter(this, 'S3AccessLoggingBucket', {
            type: 'String',
            description: 'The ARN of the access logging bucket that should be used for S3 upload bucket',
            allowedPattern: '^arn:\\S+:s3:\\S+:\\S+:*',
            constraintDescription: 'Please provide a valid ARN for the S3 access logging bucket'
        });

        const _integrationEventBus = new cdk.CfnParameter(this, 'IntegrationBus', {
            type: 'String',
            allowedPattern: '^arn:\\S+:events:\\S+:\\d{12}:event-bus/\\S+$',
            description:
                'The Event Bus to publish metadata information directly for storage and visualization. This metadata information' +
                ' does not process through the workflow engine (e.g. metadata about call analytics processing)'
        });

        const _metadataNS = new cdk.CfnParameter(this, 'MetadataNS', {
            type: 'String',
            constraintDescription: 'Please provide the namespace to used for metadata events published to event bus',
            description:
                'The namespace is used as part of the event pattern to route events to targets on the event bus'
        });

        const _stream = kinesis.Stream.fromStreamArn(this, 'PublishCommentsStream', _streamArn.valueAsString);
        const _loggingBucket = s3.Bucket.fromBucketArn(this, 'AccessLoggingBucket', _accessLoggingBucket.valueAsString);
        const _integrationBus = events.EventBus.fromEventBusArn(
            this,
            'AppIntegrationBus',
            _integrationEventBus.valueAsString
        );

        /*
        The following environment variables for lambda functiona is an example for excel based ingestion.
            let _lambdaEnv = {
                STREAM_NAME: _stream.streamName,
                INTEGRATION_BUS_NAME: _integrationBus.eventBusName,
                ID: "1",
                CREATED_DATE: "2",
                TEXT: "3",
                LANG: "4",
                ACCOUNT_NAME: "xlsx_file",
                PLATFORM: "custom"
            }
        Note the values of ID, CREATED_DATE, TEXT, and LANG are column numbers from which the value will be extracted. A sample excel file is also available
        at this path source/lambda/custom-ingestion/test/fixtures/mock_data_file.xlsx.
        */

        /*
        The following environment variables for lambda function are set up for Amazon Transcribe call analytics based ingestion of data.
        It can be used for any other JSON based ingestion as well. The source/lambda/custom-ingestion/file_processor module defines the available processor 
        types and how they can be invoked through source/lambda/custom-ingestion/file_processor/processor_factory.py
         */
        let _lambdaEnv = {
            STREAM_NAME: _stream.streamName,
            INTEGRATION_BUS_NAME: _integrationBus.eventBusName,
            PROCESSOR_TYPE: 'TRANSCRIBE_CALL_ANALYTICS',
            ID: 'GENERATE',
            CREATED_DATE: 'NOW',
            TEXT: 'Content',
            LANG: 'LanguageCode',
            ACCOUNT_NAME: 'call_analytics',
            SENTIMENT: 'Sentiment',
            NAMESPACE: _metadataNS.valueAsString,
            PLATFORM: 'customingestion',
            LIST_SELECTOR: 'Transcript'
        };

        const _s3ToEventBridgeToLambda = new S3ToEventBridgeToLambda(this, 'CustomIngestion', {
            lambdaFunctionProps: {
                runtime: lambda.Runtime.PYTHON_3_11,
                code: lambda.Code.fromAsset('lambda/ingestion-custom'),
                handler: 'lambda_function.handler',
                environment: _lambdaEnv,
                timeout: cdk.Duration.minutes(15),
                memorySize: 256
            },
            bucketProps: {
                versioned: false,
                serverAccessLogsBucket: _loggingBucket,
                serverAccessLogsPrefix: 'customingestion/'
            },
            s3LoggingBucket: _loggingBucket
        });
        this.s3Bucket = _s3ToEventBridgeToLambda.s3Bucket;

        _stream.grantWrite(_s3ToEventBridgeToLambda.lambdaFunction);
        _integrationBus.grantPutEventsTo(_s3ToEventBridgeToLambda.lambdaFunction);

        new iam.Policy(this, 'TagS3Object', {
            document: new iam.PolicyDocument({
                statements: [
                    new iam.PolicyStatement({
                        effect: iam.Effect.ALLOW,
                        actions: ['s3:PutObjectTagging'],
                        resources: [`${_s3ToEventBridgeToLambda.s3Bucket.bucketArn}/*`]
                    })
                ]
            })
        }).attachToRole(_s3ToEventBridgeToLambda.lambdaFunction.role!);

        const s3PolicyUpdateRole = new iam.Role(this, 'BucketPolicyCustomResource', {
            assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
            inlinePolicies: {
                LambdaFunctionServiceRolePolicy: new iam.PolicyDocument({
                    statements: [
                        new iam.PolicyStatement({
                            actions: ['logs:CreateLogGroup', 'logs:CreateLogStream', 'logs:PutLogEvents'],
                            resources: [
                                `arn:${cdk.Aws.PARTITION}:logs:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:log-group:/aws/lambda/*`
                            ]
                        }),
                        new iam.PolicyStatement({
                            actions: ['s3:PutBucketPolicy', 's3:GetBucketPolicy', 's3:PutBucketPublicAccessBlock'],
                            resources: [_loggingBucket.bucketArn]
                        })
                    ]
                })
            }
        });

        const s3PolicyUpdate = new lambda.Function(this, 'UpdateS3BucketPolicy', {
            handler: 'index.handler',
            runtime: lambda.Runtime.PYTHON_3_11,
            code: lambda.Code.fromInline(fs.readFileSync(path.join(__dirname, 'index.py'), 'utf8')),
            role: s3PolicyUpdateRole,
            tracing: lambda.Tracing.ACTIVE,
            timeout: cdk.Duration.minutes(15)
        });

        // prettier-ignore
        const customResourcePolicyUpdate = new cdk.CustomResource(this, 'UpdateS3Policy', {// NOSONAR - CDK construct initialization 
            resourceType: 'Custom::UpdateS3Policy',
            serviceToken: s3PolicyUpdate.functionArn,
            properties: {
                LOGGING_BUCKET_NAME: _loggingBucket.bucketName,
                SOURCE_BUCKET_NAME: this.s3Bucket.bucketName
            }
        });

        new cdk.CfnOutput(this, 'S3BucketToUploadData', {
            value: this.s3Bucket.bucketArn,
            description: 'Bucket location to upload source files for ingestion'
        });

        (s3PolicyUpdate.node.defaultChild as lambda.CfnFunction).addMetadata('cfn_nag', {
            rules_to_suppress: [
                {
                    'id': 'W89',
                    'reason': 'This is not a rule for the general case, just for specific use cases/industries'
                },
                {
                    'id': 'W92',
                    'reason': 'Impossible for us to define the correct concurrency for clients'
                }
            ]
        });

        (
            s3PolicyUpdateRole.node.tryFindChild('DefaultPolicy')!.node.tryFindChild('Resource') as iam.CfnPolicy
        ).addMetadata('cfn_nag', {
            rules_to_suppress: [
                {
                    'id': 'W12',
                    'reason':
                        'The cloudwatch policy narrows the permissions to the specific resources to the extent possible'
                }
            ]
        });
    }
}
