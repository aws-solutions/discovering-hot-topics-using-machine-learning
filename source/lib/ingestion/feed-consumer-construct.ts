#!/usr/bin/env node
/**********************************************************************************************************************
* Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
* SPDX-License-Identifier: Apache-2.0
 *********************************************************************************************************************/

import { Construct } from 'constructs';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as kinesis from 'aws-cdk-lib/aws-kinesis';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import { SqsDlq } from 'aws-cdk-lib/aws-lambda-event-sources';
import * as cdk from 'aws-cdk-lib';
import { KinesisStreamsToLambda } from '@aws-solutions-constructs/aws-kinesisstreams-lambda';
import * as defaults from '@aws-solutions-constructs/core';

export interface FeedConsumerProps {
    readonly functionProps: lambda.FunctionProps;
    readonly batchSize?: number;
    readonly retryAttempts?: number;
    readonly shardCount?: number;
    readonly retentionPeriod?: cdk.Duration;
}

export class FeedConsumer extends Construct {

    private readonly feedConsumerStream: kinesis.IStream;
    private _consumerFunction: lambda.Function;

    constructor(scope: Construct, id: string, props: FeedConsumerProps) {
        super(scope, id);

        const sqsQueue = defaults.buildQueue(this, 'SqsDlqQueue', {}).queue;

        const ingestionStream = new KinesisStreamsToLambda(this, 'IngestionStream', {
            lambdaFunctionProps: props.functionProps,
            kinesisStreamProps: {
                encryption: kinesis.StreamEncryption.MANAGED,
                retentionPeriod: props.retentionPeriod,
                shardCount: props.shardCount
            },
            kinesisEventSourceProps: {
                startingPosition: lambda.StartingPosition.TRIM_HORIZON,
                batchSize: props.batchSize,
                retryAttempts: props.retryAttempts,
                onFailure: new SqsDlq(sqsQueue)
            }
        });

        this._consumerFunction = ingestionStream.lambdaFunction;
        this.feedConsumerStream = ingestionStream.kinesisStream;

        // This dependency mapping is required because the lambda role policy is getting created before the SQS DLQ. Since the DLQ
        // does not exist the policy creation failes in CFN.
        (this._consumerFunction.role?.node.tryFindChild('DefaultPolicy') as iam.Policy).node.addDependency(sqsQueue);
    }

    public get lambdaFunction(): lambda.Function {
        return this._consumerFunction;
    }

    public get kinesisStream(): kinesis.IStream {
        return this.feedConsumerStream;
    }
}
