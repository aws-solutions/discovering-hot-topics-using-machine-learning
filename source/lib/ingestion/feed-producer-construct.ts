#!/usr/bin/env node
/**********************************************************************************************************************
 *  Copyright 2020-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                      *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICNSE-2.0                                                                     *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { AttributeType } from 'aws-cdk-lib/aws-dynamodb';
import { Rule, Schedule } from 'aws-cdk-lib/aws-events';
import { LambdaFunction } from 'aws-cdk-lib/aws-events-targets';
import { Effect, PolicyStatement, Role } from 'aws-cdk-lib/aws-iam';
import { IStream } from 'aws-cdk-lib/aws-kinesis';
import { Function, FunctionProps } from 'aws-cdk-lib/aws-lambda';
import { Aws } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { LambdaToDynamoDB } from '@aws-solutions-constructs/aws-lambda-dynamodb';

export interface FeedProducerProps {
    readonly functionProps: FunctionProps;
    readonly stream: IStream;
    readonly ingestFrequency: string;
    readonly credentialKeyPath: string;
}

export class FeedProducer extends Construct {

    private readonly producerFn: Function;

    constructor(scope: Construct, id: string, props: FeedProducerProps) {
        super(scope, id);

        const lambdaSSMPolicy = new PolicyStatement({
            effect: Effect.ALLOW,
            resources: [ `arn:${Aws.PARTITION}:ssm:${Aws.REGION}:${Aws.ACCOUNT_ID}:parameter${props.credentialKeyPath}` ],
            actions: [ 'ssm:GetParameter', 'ssm:PutParameter' ]
        });

        const dynamoDBTable = new LambdaToDynamoDB(this, 'LambdaDDB', {
            lambdaFunctionProps: props.functionProps,
            dynamoTableProps: {
                partitionKey: {
                    name: "ACCOUNT_IDENTIFIER", // socail media account identifier is the partition key
                    type: AttributeType.STRING
                },
                sortKey: {
                    name: "CREATED_TIMESTAMP",
                    type: AttributeType.STRING
                },
                timeToLiveAttribute: "EXP_DATE"
            }
        });

        props.stream.grantWrite(dynamoDBTable.lambdaFunction.role as Role)
        dynamoDBTable.lambdaFunction.addToRolePolicy(lambdaSSMPolicy);

        const rule = new Rule(this, 'PollFrequency', {
            schedule: Schedule.expression(`${props.ingestFrequency}`)
        });

        rule.addTarget(new LambdaFunction(dynamoDBTable.lambdaFunction));
        this.producerFn = dynamoDBTable.lambdaFunction;
    }

    public get producerFunction(): Function {
        return this.producerFn;
    }
}
