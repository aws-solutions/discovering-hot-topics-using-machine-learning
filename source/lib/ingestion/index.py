import json
import logging
import os
import uuid

import boto3
import cfnresponse

RESOURCE_PROPERTIES = "ResourceProperties"
SOURCE_BUCKET_NAME = "SOURCE_BUCKET_NAME"
LOGGING_BUCKET_NAME = "LOGGING_BUCKET_NAME"
PHYSICAL_RESOURCE_ID = "PhysicalResourceId"
POLICY_SID = "customingestionAccessLogsPolicy"
STATEMENT = "Statement"

DEFAULT_LEVEL = "DEBUG"

s3 = boto3.client("s3")
logger = logging.getLogger(__name__)
logger.setLevel(os.environ.get("LOG_LEVEL", DEFAULT_LEVEL))


def handler(event, context):
    logger.debug(f"The event is {event}")
    physical_resource_id = None

    try:
        physical_resource_id = event.get(PHYSICAL_RESOURCE_ID, uuid.uuid4().hex[:8])

        if event["RequestType"] == "Create" or event["RequestType"] == "Update":
            source_bucket_name = event[RESOURCE_PROPERTIES][SOURCE_BUCKET_NAME]
            logging_bucket_name = event[RESOURCE_PROPERTIES][LOGGING_BUCKET_NAME]
            account_id = context.invoked_function_arn.split(":")[4]
            aws_partition = context.invoked_function_arn.split(":")[1]

            s3.put_public_access_block(
                Bucket=logging_bucket_name,
                PublicAccessBlockConfiguration={
                    "BlockPublicAcls": True,
                    "IgnorePublicAcls": True,
                    "BlockPublicPolicy": False,
                    "RestrictPublicBuckets": True,
                },
                ExpectedBucketOwner=account_id,
            )

            policy_string = s3.get_bucket_policy(Bucket=logging_bucket_name)["Policy"]
            bucket_policy = json.loads(policy_string)

            update_policy = True

            for statement in bucket_policy[STATEMENT]:
                logger.debug(f"Statement is {json.dumps(statement)}")
                if "Sid" in statement and statement["Sid"] == POLICY_SID:
                    if (
                        f"arn:{aws_partition}:s3:::{source_bucket_name}"
                        not in statement["Condition"]["ArnLike"]["aws:SourceArn"]
                    ):
                        logger.info("Bucket policy exists but with a different bucket, hence removing it")
                        bucket_policy[STATEMENT].remove(statement)
                        break
                    else:
                        logger.info("Bucket policy exists, hence update not required")
                        update_policy = False
                        break

            if update_policy:
                bucket_policy[STATEMENT].append(
                    {
                        "Action": "s3:PutObject",
                        "Condition": {
                            "ArnLike": {"aws:SourceArn": [f"arn:{aws_partition}:s3:::{source_bucket_name}"]},
                            "StringEquals": {"aws:SourceAccount": account_id},
                        },
                        "Effect": "Allow",
                        "Principal": {"Service": "logging.s3.amazonaws.com"},
                        "Resource": f"arn:{aws_partition}:s3:::{logging_bucket_name}/customingestion*",
                        "Sid": POLICY_SID,
                    }
                )

                s3.put_bucket_policy(Bucket=logging_bucket_name, Policy=json.dumps(bucket_policy))

            s3.put_public_access_block(
                Bucket=logging_bucket_name,
                PublicAccessBlockConfiguration={
                    "BlockPublicAcls": True,
                    "IgnorePublicAcls": True,
                    "BlockPublicPolicy": True,
                    "RestrictPublicBuckets": True,
                },
                ExpectedBucketOwner=account_id,
            )
            logger.info("Policy Updated successfully. Sending success response to CloudFormation")

        cfnresponse.send(event, context, cfnresponse.SUCCESS, {}, physical_resource_id)
    except Exception as ex:
        logger.error(f"Failed to update policy, error is {str(ex)}")
        cfnresponse.send(
            event, context, cfnresponse.FAILED, {}, physicalResourceId=physical_resource_id, reason=str(ex)
        )
