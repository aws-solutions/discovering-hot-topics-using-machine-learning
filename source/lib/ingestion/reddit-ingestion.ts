#!/usr/bin/env node
/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import * as ddb from '@aws-cdk/aws-dynamodb';
import * as events from '@aws-cdk/aws-events';
import * as kinesis from '@aws-cdk/aws-kinesis';
import * as lambda from '@aws-cdk/aws-lambda';
import * as targets from '@aws-cdk/aws-events-targets';
import * as cdk from '@aws-cdk/core';
import { DiscoveringHotTopicsStack } from '../discovering-hot-topics-stack';
import { DataIngestionTemplate } from './data-ingestion-template';

export class RedditIngestion extends cdk.NestedStack {
    readonly reddit_namespace: string = 'com.analyze.reddit.subreddit';

    constructor(scope: cdk.Construct, id: string, props: cdk.NestedStackProps) {
        super(scope, id, props);

        const _redditAPIKey = new cdk.CfnParameter(this, 'RedditAPIKey', {
            type: 'String',
            description:
                'Required: The SSM parameter key name where the Reddit API credentials detailare stored. For ' +
                'details about how and where to store the API credentials, please refer the implementation guide for this solution',
            default: '/discovering-hot-topics-using-machine-learning/reddit/comments',
            constraintDescription: 'Please provide the SSM key for Reddit API'
        });

        const _eventBusArn = new cdk.CfnParameter(this, 'EventBus', {
            type: 'String',
            description: 'Required: The ARN of the Event Bus messaging backbone',
            allowedPattern: '^arn:\\S+:events:\\S+:\\d{12}:event-bus/\\S+$',
            constraintDescription: 'Please provide the ARN of the Event Bus messaging backbone'
        });

        const _streamArn = new cdk.CfnParameter(this, 'StreamARN', {
            type: 'String',
            description: 'Required: The name of the stream where search comments should be published for analysis',
            allowedPattern: '^arn:\\S+:kinesis:\\S+:\\d{12}:stream/\\S+$',
            constraintDescription: 'Please provide the Amazon Kinesis Data Streams name'
        });

        const _subRedditIngestionFreq = new cdk.CfnParameter(this, 'RedditIngestionFrequency', {
            type: 'String',
            description:
                'Required: The Polling frequency at which the system should ingestion comments from subreddits',
            default: 'cron(0/60 * * * ? *)',
            allowedPattern: `${DiscoveringHotTopicsStack.cronRegex}`
        });

        const _subRedditsToFollow = new cdk.CfnParameter(this, 'SubRedditsToFollow', {
            type: 'String',
            description:
                'Optional: Please provide the list of SubReddits to follow as comma separated list. Alternatively you ' +
                'can also set the list in the DynamoDB table. For details on the DynamoDB configuration, please refer our implementation guide',
            constraintDescription: '',
            allowedPattern: '[r\\/\\w+]+([,r\\/\\w+])*',
            default: 'r/aws,r/MachineLearning'
        });

        const _stream = kinesis.Stream.fromStreamArn(this, 'PublishCommentsStream', _streamArn.valueAsString);
        const _eventBus = events.EventBus.fromEventBusArn(this, 'Bus', _eventBusArn.valueAsString) as events.EventBus;
        const _solution_id = this.node.tryGetContext('solution_id');
        const _solution_version = this.node.tryGetContext('solution_version');

        const _redditDataIngestion = new DataIngestionTemplate(this, 'SubReddit', {
            source: {
                lambdaFunctionProps: {
                    description: 'Lambda function to publish the subreddits to ingest information from',
                    handler: 'publish-subreddit.handler',
                    code: lambda.Code.fromAsset('lambda/ingestion-reddit'),
                    runtime: lambda.Runtime.NODEJS_14_X,
                    environment: {
                        SUBREDDITS_TO_FOLLOW: _subRedditsToFollow.valueAsString,
                        SUBREDDIT_PUBLISH_NAMESPACE: this.reddit_namespace
                    },
                    timeout: cdk.Duration.minutes(15)
                },
                tableProps: {
                    partitionKey: {
                        name: 'SUB_REDDIT',
                        type: ddb.AttributeType.STRING
                    }
                }
            },
            target: {
                lambdaFunctionProps: {
                    description: 'Lambda function to ingest comments from subreddits of interest',
                    handler: 'subreddit-comment.handler',
                    code: lambda.Code.fromAsset('lambda/ingestion-reddit'),
                    runtime: lambda.Runtime.NODEJS_14_X,
                    environment: {
                        REDDIT_API_KEY: _redditAPIKey.valueAsString,
                        SOLUTION_VERSION: _solution_version,
                        SOLUTION_ID: _solution_id,
                        STACK_NAME: cdk.Aws.STACK_NAME,
                        STREAM_NAME: _stream.streamName
                    },
                    timeout: cdk.Duration.minutes(15),
                    reservedConcurrentExecutions: 1 // setting to 1 to have control on API throttling from Reddit
                },
                tableProps: {
                    partitionKey: {
                        name: 'SUB_REDDIT',
                        type: ddb.AttributeType.STRING
                    }
                },
                credentialKeyPath: _redditAPIKey.valueAsString
            },
            ingestionEventRuleProps: {
                eventPattern: {
                    account: [cdk.Aws.ACCOUNT_ID],
                    region: [cdk.Aws.REGION],
                    source: [this.reddit_namespace]
                }
            },
            existingIngestionEventBus: _eventBus
        });
        _stream.grantWrite(_redditDataIngestion.targetLambda);

        const rule = new events.Rule(this, 'PollFrequency', {
            schedule: events.Schedule.expression(_subRedditIngestionFreq.valueAsString)
        });

        rule.addTarget(new targets.LambdaFunction(_redditDataIngestion.sourceLambda));
    }
}
