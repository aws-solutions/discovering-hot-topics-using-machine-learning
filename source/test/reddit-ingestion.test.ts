/**********************************************************************************************************************
* Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
* SPDX-License-Identifier: Apache-2.0
 *********************************************************************************************************************/

import { Template } from 'aws-cdk-lib/assertions';
import * as events from 'aws-cdk-lib/aws-events';
import * as kinesis from 'aws-cdk-lib/aws-kinesis';
import * as cdk from 'aws-cdk-lib';
import { RedditIngestion } from '../lib/ingestion/reddit-ingestion';

test('test reddit ingestion stack', () => {
    const stack = new cdk.Stack();
    const _eventBus = new events.EventBus(stack, 'Bus');
    const _stream = new kinesis.Stream(stack, 'Stream', {});

    new RedditIngestion(stack, 'testredditingestion', {
        parameters: {
            RedditAPIKey: '/fakelocation/reddit/comments',
            EventBus: _eventBus.eventBusArn,
            StreamARN: _stream.streamArn,
            RedditIngestionFrequency: 'cron(0/60 * * * ? *)',
            SubRedditsToFollow: 'r/test1,r/test2'
        }
    });
});
