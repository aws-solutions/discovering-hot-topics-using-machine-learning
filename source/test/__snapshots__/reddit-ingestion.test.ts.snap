// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test reddit ingestion stack 1`] = `
{
  "Conditions": {
    "AwsCdkKinesisEncryptedStreamsUnsupportedRegions": {
      "Fn::Or": [
        {
          "Fn::Equals": [
            {
              "Ref": "AWS::Region",
            },
            "cn-north-1",
          ],
        },
        {
          "Fn::Equals": [
            {
              "Ref": "AWS::Region",
            },
            "cn-northwest-1",
          ],
        },
      ],
    },
  },
  "Parameters": {
    "AssetParameters054cb1e445342f40cca00b9a91cd16d3515c346e6d8c2c2fcc90df1287112fcaArtifactHashB57AF0DB": {
      "Description": "Artifact hash for asset "054cb1e445342f40cca00b9a91cd16d3515c346e6d8c2c2fcc90df1287112fca"",
      "Type": "String",
    },
    "AssetParameters054cb1e445342f40cca00b9a91cd16d3515c346e6d8c2c2fcc90df1287112fcaS3Bucket2AC17F32": {
      "Description": "S3 bucket for asset "054cb1e445342f40cca00b9a91cd16d3515c346e6d8c2c2fcc90df1287112fca"",
      "Type": "String",
    },
    "AssetParameters054cb1e445342f40cca00b9a91cd16d3515c346e6d8c2c2fcc90df1287112fcaS3VersionKey5C9A8948": {
      "Description": "S3 key for asset version "054cb1e445342f40cca00b9a91cd16d3515c346e6d8c2c2fcc90df1287112fca"",
      "Type": "String",
    },
    "AssetParametersa9398e7cb0666fe22137031df2cb407da83ec58860a26c9fd2460acc940df504ArtifactHash1E6D23A4": {
      "Description": "Artifact hash for asset "a9398e7cb0666fe22137031df2cb407da83ec58860a26c9fd2460acc940df504"",
      "Type": "String",
    },
    "AssetParametersa9398e7cb0666fe22137031df2cb407da83ec58860a26c9fd2460acc940df504S3Bucket1CD04397": {
      "Description": "S3 bucket for asset "a9398e7cb0666fe22137031df2cb407da83ec58860a26c9fd2460acc940df504"",
      "Type": "String",
    },
    "AssetParametersa9398e7cb0666fe22137031df2cb407da83ec58860a26c9fd2460acc940df504S3VersionKey7872D359": {
      "Description": "S3 key for asset version "a9398e7cb0666fe22137031df2cb407da83ec58860a26c9fd2460acc940df504"",
      "Type": "String",
    },
  },
  "Resources": {
    "BusEA82B648": {
      "Properties": {
        "Name": "Bus",
      },
      "Type": "AWS::Events::EventBus",
    },
    "Stream790BDEE4": {
      "Properties": {
        "RetentionPeriodHours": 24,
        "ShardCount": 1,
        "StreamEncryption": {
          "Fn::If": [
            "AwsCdkKinesisEncryptedStreamsUnsupportedRegions",
            {
              "Ref": "AWS::NoValue",
            },
            {
              "EncryptionType": "KMS",
              "KeyId": "alias/aws/kinesis",
            },
          ],
        },
        "StreamModeDetails": {
          "StreamMode": "PROVISIONED",
        },
      },
      "Type": "AWS::Kinesis::Stream",
    },
    "testredditingestionNestedStacktestredditingestionNestedStackResourceE8B6AADB": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "Parameters": {
          "EventBus": {
            "Fn::GetAtt": [
              "BusEA82B648",
              "Arn",
            ],
          },
          "RedditAPIKey": "/fakelocation/reddit/comments",
          "RedditIngestionFrequency": "cron(0/60 * * * ? *)",
          "StreamARN": {
            "Fn::GetAtt": [
              "Stream790BDEE4",
              "Arn",
            ],
          },
          "SubRedditsToFollow": "r/test1,r/test2",
          "referencetoAssetParameters054cb1e445342f40cca00b9a91cd16d3515c346e6d8c2c2fcc90df1287112fcaS3Bucket2AC17F32Ref": {
            "Ref": "AssetParameters054cb1e445342f40cca00b9a91cd16d3515c346e6d8c2c2fcc90df1287112fcaS3Bucket2AC17F32",
          },
          "referencetoAssetParameters054cb1e445342f40cca00b9a91cd16d3515c346e6d8c2c2fcc90df1287112fcaS3VersionKey5C9A8948Ref": {
            "Ref": "AssetParameters054cb1e445342f40cca00b9a91cd16d3515c346e6d8c2c2fcc90df1287112fcaS3VersionKey5C9A8948",
          },
        },
        "TemplateURL": {
          "Fn::Join": [
            "",
            [
              "https://s3.",
              {
                "Ref": "AWS::Region",
              },
              ".",
              {
                "Ref": "AWS::URLSuffix",
              },
              "/",
              {
                "Ref": "AssetParametersa9398e7cb0666fe22137031df2cb407da83ec58860a26c9fd2460acc940df504S3Bucket1CD04397",
              },
              "/",
              {
                "Fn::Select": [
                  0,
                  {
                    "Fn::Split": [
                      "||",
                      {
                        "Ref": "AssetParametersa9398e7cb0666fe22137031df2cb407da83ec58860a26c9fd2460acc940df504S3VersionKey7872D359",
                      },
                    ],
                  },
                ],
              },
              {
                "Fn::Select": [
                  1,
                  {
                    "Fn::Split": [
                      "||",
                      {
                        "Ref": "AssetParametersa9398e7cb0666fe22137031df2cb407da83ec58860a26c9fd2460acc940df504S3VersionKey7872D359",
                      },
                    ],
                  },
                ],
              },
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;
